plugins/by-name/typescript-tools/default.nix

```diff
@@ -157,38 +157,43 @@ in
       };
     };
   };
-  config = mkIf cfg.enable {
-    extraPlugins = [ cfg.package ];
 
-    plugins.lsp.postConfig =
-      with cfg;
+  # NOTE: call setup in lsp config
+  callSetup = false;
+  extraConfig = cfg: {
+    plugins.lsp.postConfig = ''
+      require('typescript-tools').setup(${lib.nixvim.toLuaObject cfg.settings})
+    '';
+
+    warnings =
       let
-        options = {
-          inherit handlers;
-          on_attach = onAttach;
-
-          settings = with settings; {
-            separate_diagnostic_server = separateDiagnosticServer;
-            publish_diagnostic_on = publishDiagnosticOn;
-            expose_as_code_action = exposeAsCodeAction;
-            tsserver_path = tsserverPath;
-            tsserver_plugins = tsserverPlugins;
-            tsserver_max_memory = tsserverMaxMemory;
-            tsserver_format_options = tsserverFormatOptions;
-            tsserver_file_preferences = tsserverFilePreferences;
-            tsserver_locale = tsserverLocale;
-            complete_function_calls = completeFunctionCalls;
-            include_completions_with_insert_text = includeCompletionsWithInsertText;
-            code_lens = codeLens;
-            disable_member_code_lens = disableMemberCodeLens;
-            jsx_close_tag = with jsxCloseTag; {
-              inherit enable filetypes;
-            };
-          };
-        };
+        definedOpts = lib.filter (opt: lib.hasAttrByPath (lib.toList opt) cfg.settings) [
+          "separateDiagnosticServer"
+          "publishDiagnosticOn"
+          "exposeAsCodeAction"
+          "tsserverPath"
+          "tsserverPlugins"
+          "tsserverMaxMemory"
+          "tsserverFormatOptions"
+          "tsserverFilePreferences"
+          "tsserverLocale"
+          "completeFunctionCalls"
+          "includeCompletionsWithInsertText"
+          "codeLens"
+          "disableMemberCodeLens"
+          [
+            "jsxCloseTag"
+            "enable"
+          ]
+          [
+            "jsxCloseTag"
+            "filetypes"
+          ]
+        ];
```

Comment:
This would benefit from a) a "TODO: added; remove after" comment and b) some explanatory comments regarding why we can't just do a normal optionRenamedToSettings